bakery/codec.go

```diff
@@ -23,109 +28,158 @@ type caveatId struct {
 	Id                  string
 }
 
-// boxEncoder encodes caveat ids confidentially to a third-party service using
-// authenticated public key encryption compatible with NaCl box.
-type boxEncoder struct {
-	key *KeyPair
-}
-
-// newBoxEncoder creates a new boxEncoder that uses the given public key pair.
-func newBoxEncoder(key *KeyPair) *boxEncoder {
-	return &boxEncoder{
-		key: key,
-	}
-}
-
-func (enc *boxEncoder) encodeCaveatId(condition string, rootKey []byte, thirdPartyPub *PublicKey) (string, error) {
-	id, err := enc.newCaveatId(condition, rootKey, thirdPartyPub)
-	if err != nil {
-		return "", err
-	}
-	data, err := json.Marshal(id)
-	if err != nil {
-		return "", fmt.Errorf("cannot marshal %#v: %v", id, err)
-	}
-	return base64.StdEncoding.EncodeToString(data), nil
-}
-
-func (enc *boxEncoder) newCaveatId(condition string, rootKey []byte, thirdPartyPub *PublicKey) (*caveatId, error) {
+// encodeJSONCaveatId creates a JSON encoded third-party caveat.
+func encodeJSONCaveatId(key *KeyPair, ci caveatInfo) ([]byte, error) {
 	var nonce [NonceLen]byte
 	if _, err := rand.Read(nonce[:]); err != nil {
-		return nil, fmt.Errorf("cannot generate random number for nonce: %v", err)
+		return nil, errgo.Notef(err, "cannot generate random number for nonce")
 	}
 	plain := caveatIdRecord{
-		RootKey:   rootKey,
-		Condition: condition,
+		RootKey:   ci.RootKey,
+		Condition: ci.Predicate,
 	}
 	plainData, err := json.Marshal(&plain)
 	if err != nil {
-		return nil, fmt.Errorf("cannot marshal %#v: %v", &plain, err)
+		return nil, errgo.Notef(err, "cannot marshal %#v", &plain)
 	}
-	sealed := box.Seal(nil, plainData, &nonce, thirdPartyPub.boxKey(), enc.key.Private.boxKey())
-	return &caveatId{
-		ThirdPartyPublicKey: thirdPartyPub,
-		FirstPartyPublicKey: &enc.key.Public,
+	sealed := key.seal(ci.PeerPublicKey, &nonce, plainData)
+	id := caveatId{
+		ThirdPartyPublicKey: ci.PeerPublicKey,
+		FirstPartyPublicKey: &key.Public,
 		Nonce:               nonce[:],
 		Id:                  base64.StdEncoding.EncodeToString(sealed),
-	}, nil
+	}
+	data, err := json.Marshal(id)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot marshal %#v", id)
+	}
+	buf := make([]byte, base64.StdEncoding.EncodedLen(len(data)))
+	base64.StdEncoding.Encode(buf, data)
+	return buf, nil
 }
 
-// boxDecoder decodes caveat ids for third-party service that were encoded to
-// the third-party with authenticated public key encryption compatible with
-// NaCl box.
-type boxDecoder struct {
-	key *KeyPair
+const (
+	v0ThirdPartyPublicKeyPrefixOffset = 1
+	v0FirstPartyPublicKeyOffset       = v0ThirdPartyPublicKeyPrefixOffset + 4
+	v0NonceOffset                     = v0FirstPartyPublicKeyOffset + KeyLen
+	v0EncryptedPartOffset             = v0NonceOffset + NonceLen
+
+	v0RootKeyOffset    = 1
+	v0CaveatTextOffset = v0RootKeyOffset + 24
+)
+
+// encodeCaveatIdV0 creates a version 0 third-party caveat.
```

Comment:
I'd add a comment describing the format:

// Format:
// 
// version 0 [1 byte]
// first 4 bytes of third-party Curve25519 public key [4 bytes]
// first-party Curve25519 public key [32 bytes]
// nonce [24 bytes]
// encrypted secret part [rest of message]